### PR TITLE
Fix and clean up how a merged domain is determined

### PR DESCRIPTION
**Pull request by naoyam**
Background:
- https://github.com/csarofeen/pytorch/pull/731/files#r617773798
- https://github.com/csarofeen/pytorch/pull/731/files#r617774948

There's actually a bug in the code. `getVectorizeExtent` is supposed to be called with a global TV, but wasn't.

It uses the contiguity information rather than looking at indices. Also, extending the merged domain does not go beyond the CA position.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 1]
```diff
@@ -124,40 +124,43 @@ kir::Expr* findVectorizedSet(
 
 // Get full extent for the inner-most, merged root domain
 kir::Val* getVectorizeExtent(
-    kir::TensorView* tv,
-    const std::vector<kir::Val*>& indices) {
-  auto domain = tv->domain()->hasRFactor() ? tv->domain()->rfactorDomain()
-                                           : tv->domain()->rootDomain();
+    kir::TensorView* global_tv,
+    kir::TensorView* local_tv) {
+  TORCH_INTERNAL_ASSERT(!global_tv->domain()->hasRFactor());
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Should we check that `global_tv` has valid contiguity information?
1. check `global_tv` is in global memory
2. check `global_tv` is either an input or output to the fusion

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp [hunk 2]
```diff
@@ -124,40 +124,43 @@ kir::Expr* findVectorizedSet(
...
-    auto root_id = domain.at(i);
+  for (int i = int(root_domain.size()) - 1; i >= 0; --i) {
+    auto root_id = root_domain.at(i);
     if (root_id->isBroadcast() || root_id->isReduction() ||
         gpu_lower->trivialReductionInfo().isDerived(root_id)) {
       continue;
-    } else if (extent == nullptr) {
+    }
+    // Don't extend the vectorization domain beyond the CA position
+    if (root_right_of_ca_domains.find(local_fuser_tv->getRootDomain()[i]) ==
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Would using a `PairwiseRootDomainMap` make sense here to ensure a correct mapping between `global_tv` and `local_tv`?

**Review by rdspring1 — APPROVED** [review]
LGTM. I just left some comments.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: That's a good point. I'll fix it.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Can contiguity information be invalid?

**Comment by naoyam** [comment]
@rdspring1 I updated `getVectorizeExtent` with more comprehensive checking. I *believe* it covers all potential cases, but please let me know if you find any failing case.

**Comment by rdspring1** [comment]
LGTM

**Pull request merged**